### PR TITLE
Fix 2222 - improve output of the outdated warning and fix underlying bug

### PR DESCRIPTION
**Pull request by matthid**
fixes both issues in #2222.

Before
```
paket.dependencies and paket.lock are out of sync in C:\PROJ\FAKE.
Please run 'paket install' or 'paket update' to recompute the paket.lock file.
```

After
```
paket.dependencies and paket.lock are out of sync in C:\PROJ\FAKE.
Please run 'paket install' or 'paket update' to recompute the paket.lock file.
Changes [RestrictionsChanged] were detected in netcore/FSharp.Compiler.Service
Changes [RestrictionsChanged] were detected in netcore/FSharp.Core
Changes [RestrictionsChanged] were detected in netcore/Libuv
Changes [RestrictionsChanged] were detected in netcore/Microsoft.CodeAnalysis.Common
Changes [RestrictionsChanged] were detected in netcore/Microsoft.CodeAnalysis.CSharp
Changes [RestrictionsChanged] were detected in netcore/Microsoft.CodeAnalysis.VisualBasic
Changes [RestrictionsChanged] were detected in netcore/Microsoft.CSharp
Changes [RestrictionsChanged] were detected in netcore/Microsoft.DiaSymReader
Changes [RestrictionsChanged] were detected in netcore/Microsoft.DiaSymReader.PortablePdb
Changes [RestrictionsChanged] were detected in netcore/Microsoft.NETCore.App
Changes [RestrictionsChanged] were detected in netcore/Microsoft.NETCore.DotNetHost
Changes [RestrictionsChanged] were detected in netcore/Microsoft.NETCore.DotNetHostPolicy
Changes [RestrictionsChanged] were detected in netcore/Microsoft.NETCore.DotNetHostResolver
Changes [RestrictionsChanged] were detected in netcore/Microsoft.NETCore.Jit
Changes [RestrictionsChan...
```

### CODE REVIEW

**Comment by matthid** [comment]
apparently the problem was that paket could not properly parse `netcore10` (which it wrote itself).

**Comment by matthid** [comment]
Lets see what the CI has to say :)

File: src/Paket.Core/PackageResolver.fs [hunk 1]
```diff
@@ -651,7 +651,7 @@ let Resolve (getVersionsF, getPackageDetailsF, groupName:GroupName, globalStrate
 
     let rec step (stage:Stage) (stackpack:StackPack) compatibleVersions (flags:StepFlags) =
 
-        let fuseConflicts currentConflict priorConflictSteps =
+        let inline fuseConflicts currentConflict priorConflictSteps =
```
Comment (matthid) [inline src/Paket.Core/PackageResolver.fs]: @cloudRoutine Is this OK? This prevents a stack-overflow I had when debugging (compiling in DEBUG mode)

File: src/Paket.Core/FrameworkHandling.fs [hunk 2]
```diff
@@ -365,8 +365,8 @@ module FrameworkDetection =
                 | "netstandard14" -> Some(DotNetStandard DotNetStandardVersion.V1_4)
                 | "netstandard15" -> Some(DotNetStandard DotNetStandardVersion.V1_5)
                 | "netstandard16" -> Some(DotNetStandard DotNetStandardVersion.V1_6)
-                | "netcore10" | "netcoreapp10" -> Some (DotNetCore DotNetCoreVersion.V1_0)
-                | "netcore11" | "netcoreapp11" -> Some (DotNetCore DotNetCoreVersion.V1_1)
+                | "netcore10" -> Some (DotNetCore DotNetCoreVersion.V1_0)
+                | "netcore11" -> Some (DotNetCore DotNetCoreVersion.V1_1)
```
Comment (matthid) [inline src/Paket.Core/FrameworkHandling.fs]: Note: This is in fact cleanup as we have the `netcoreapp -> netcore` alias defined above.

File: src/Paket.Core/RestoreProcess.fs [hunk 3]
```diff
@@ -193,11 +193,13 @@ let Restore(dependenciesFileName,projectFile,force,group,referencesFileNames,ign
...
 
     if not hasLocalFile && not ignoreChecks then
-        let hasAnyChanges,_,_,_ = DependencyChangeDetection.GetChanges(dependenciesFile,lockFile,false)
+        let hasAnyChanges,nugetChanges,remoteFilechanges,hasChanges = DependencyChangeDetection.GetChanges(dependenciesFile,lockFile,false)
 
         let checkResponse = if failOnChecks then failwithf else traceWarnfn
         if hasAnyChanges then 
             checkResponse "paket.dependencies and paket.lock are out of sync in %s.%sPlease run 'paket install' or 'paket update' to recompute the paket.lock file." lockFileName.Directory.FullName Environment.NewLine
+            for (group, package, changes) in nugetChanges do
+                traceWarnfn "Changes %A were detected in %s/%s" changes (group.ToString()) (package.ToString())
```
Comment (matthid) [inline src/Paket.Core/RestoreProcess.fs]: Generating a really nice output is left as an exercise to the reader :)

**Comment by matthid** [comment]
OK This is ready. The AppVeyor error is unrelated and on the master branch as well.

File: src/Paket.Core/PackageResolver.fs (hunk 1 above)
Comment (forki) [inline src/Paket.Core/PackageResolver.fs]: interesting. probably just because the stack size was halved due to inlining. but I'm fine with that!

File: src/Paket.Core/RestoreProcess.fs (hunk 3 above)
Comment (forki) [inline src/Paket.Core/RestoreProcess.fs]: yeah we can iterate on that

File: src/Paket/Paket.fsproj [hunk 4]
```diff
@@ -37,8 +37,8 @@
     <StartWorkingDirectory>D:\temp\coreclrtest</StartWorkingDirectory>
     <StartArguments>install</StartArguments>
     <StartWorkingDirectory>C:\dev\src\Paket\integrationtests\scenarios\loading-scripts\dependencies-file-flag\temp</StartWorkingDirectory>
-    <StartArguments>install --createnewbindingfiles</StartArguments>
-    <StartWorkingDirectory>D:\temp\pakkit2</StartWorkingDirectory>
+    <StartArguments>install</StartArguments>
```
Comment (forki) [inline src/Paket/Paket.fsproj]: I wished this would be in different file so that we could gitignore.

**Comment by forki** [comment]
Good work! Thanks a lot

File: src/Paket.Core/PackageResolver.fs (hunk 1 above)
Comment (matthid) [inline src/Paket.Core/PackageResolver.fs]: Ah okey, I thought the thing is now tail recursive but the compiler is not picking it up because of this in debug mode. Thanks for the clarification.

File: src/Paket/Paket.fsproj (hunk 4 above)
Comment (matthid) [inline src/Paket/Paket.fsproj]: I can remove that if you want.

Comment (matthid) [inline src/Paket/Paket.fsproj]: Maybe after @Krzysztof-Cieslak brings the toml file ;)

Comment (forki) [inline src/Paket/Paket.fsproj]: no it's fine. It's more a general observation. I'm really annoyed about that part of fsproj

Comment (forki) [inline src/Paket/Paket.fsproj]: yeah not sure if MSbuild 15 is better in that regard. but VS can't read it anyways ;-)

Comment (matthid) [inline src/Paket/Paket.fsproj]: Well maybe we soon use all VSCode with ionide (when the full .net debugger works) and we don't have this problem anymore (because VSCode uses its own json files for debugging?)

File: src/Paket.Core/Requirements.fs [hunk 5]
```diff
@@ -60,13 +70,17 @@ let parseRestrictions(text:string) =
...
             if operatorSplit.[0] = ">=" then
                 if operatorSplit.Length < 4 then
                     yield FrameworkRestriction.AtLeast x
                 else
-                    match FrameworkDetection.Extract(operatorSplit.[3]) with
-                    | None -> ()
+                    let item = operatorSplit.[3]
+                    match FrameworkDetection.Extract(item) with
+                    | None ->
+                        handleError <| sprintf "Could not parse second framework of between operator '%s'. Try to update or install again or report a paket bug." item
```
Comment (matthid) [inline src/Paket.Core/Requirements.fs]: @forki Just in case I cannot react fast enough and people are experiencing failures tomorrow: We can just set `failImmediatly` to `false` within the `parseRestrictions` function (as first operation, essentially ignoring the parameter). This way paket will log errors instead of failing directly. But more often than not this indicates a bug somewhere...

Comment (forki) [inline src/Paket.Core/Requirements.fs]: Good thanks for the warning.

Comment (forki) [inline src/Paket.Core/Requirements.fs]: Maybe I should revert the change, push a bug fix version with old code and publish this as minor version change?

Comment (matthid) [inline src/Paket.Core/Requirements.fs]: I don't think this is a new feature, I have not met a case where this actually triggered wrongly.

I encountered two cases: 
 - The file was written by paket and can no longer be parsed -> paket bug (another minor increase)
 - The file was written by the user (paket.references for example in my test case) -> user error as he entered some invalid value, its just that paket never told him.

All in all I think a patch version is enough as this indeed fixes a bug (one could argue if the additional output is a new feature though...)

Comment (forki) [inline src/Paket.Core/Requirements.fs]: doing it anyways. just in case.

Comment (matthid) [inline src/Paket.Core/Requirements.fs]: So if something belongs into a minor release its the logic required to produce the additional output and not this part ;). But that's of IMHO.

**Comment by matthid** [comment]
Finally the warning is gone. As always thanks for merging and releasing so quickly :)
